src/main/kotlin/org/rust/ide/annotator/cargoCheck/RsCargoCheckAnnotator.kt

```diff
@@ -18,52 +18,90 @@ import com.intellij.openapi.editor.Editor
 import org.rust.lang.core.psi.RsFile
-import org.rust.lang.core.psi.ext.cargoWorkspace
+import org.rust.lang.core.psi.ext.cargoProject
 import org.rust.lang.core.psi.ext.containingCargoPackage
+import org.rust.openapiext.checkReadAccessAllowed
+import org.rust.openapiext.checkReadAccessNotAllowed
 import org.rust.openapiext.isUnitTestMode
 import org.rust.openapiext.saveAllDocuments
 import java.lang.reflect.Field
 
-object RsCargoCheckAnnotator : ExternalAnnotator<RsCargoCheckAnnotationInfo, RsCargoCheckAnnotationResult>() {
+object RsCargoCheckAnnotator : ExternalAnnotator<Lazy<RsCargoCheckAnnotationResult?>, RsCargoCheckAnnotationResult>() {
     private val LOG: Logger = Logger.getInstance(RsCargoCheckAnnotator::class.java)
     const val TEST_MESSAGE: String = "CargoCheckAnnotation"
 
-    override fun collectInformation(file: PsiFile): RsCargoCheckAnnotationInfo? {
+    override fun collectInformation(file: PsiFile): Lazy<RsCargoCheckAnnotationResult?>? {
+        checkReadAccessAllowed()
         if (file !is RsFile) return null
         if (!file.project.rustSettings.useCargoCheckAnnotator) return null
-        val ws = file.cargoWorkspace ?: return null
-        val module = ModuleUtil.findModuleForFile(file.virtualFile, file.project) ?: return null
-        val toolchain = module.project.toolchain ?: return null
-        val cargoPackage = file.containingCargoPackage
-        if (cargoPackage?.origin != PackageOrigin.WORKSPACE) return null
-        return RsCargoCheckAnnotationInfo(toolchain, ws.contentRoot, module, cargoPackage)
+
+        val cargoProject = file.cargoProject ?: return null
+        if (cargoProject.project.toolchain == null) return null
+
+        return CachedValuesManager.getManager(file.project)
+            .getCachedValue(cargoProject) {
+                // We want to run `cargo check` in background thread and *without* read action.
+                // And we also want to cache results of `cargo check` because `cargo check` is package-global,
+                // but annotator can be invoked separately for each file.
+                // With CachedValuesManager our cached value should be invalidated on any PSI change.
+                // Important note about this cache is that modification count will be stored AFTER computation
+                // of a value. If we aren't in read action, PSI can be modified during computation of the value
+                // and so an outdated value will be cached. So we can't use the cache without read action.
+                // What we really want:
+                // 1. store current psi modification count
+                // 2. run "cargo check" and retrieve results (in background thread and without read action)
+                // 3. try to cache results use modification count stored in (1). Results can be already outdated here
+                // We got such behavior by this hack: we store "Lazy" computation to the cache. We do it in read
```

Comment:
Maybe the following variant is a little bit better?
"We get such behavior by storing `Lazy` computation to the cache. Cache result is created in read action, so it will be stored within correct PSI modification count. Then, we will retrieve the value from Lazy in a background thread. The value will be computed or retrieved from the already computed `Lazy` value."